File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/network/JsonAnswerSerializer.java
package com.github.dedis.popstellar.model.network.serializer.network;

import com.github.dedis.popstellar.model.network.answer.Error;
import com.github.dedis.popstellar.model.network.answer.*;
import com.github.dedis.popstellar.model.network.serializer.JsonUtils;
import com.google.gson.*;

import java.lang.reflect.Type;

/** Json serializer and deserializer for the answer */
public class JsonAnswerSerializer implements JsonSerializer<Answer>, JsonDeserializer<Answer> {

  private static final String RESULT = "result";
  private static final String ERROR = "error";

  @Override
  public Answer deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    JsonObject obj = json.getAsJsonObject();
    JsonUtils.testRPCVersion(obj);
    JsonUtils.verifyJson(JsonUtils.ROOT_SCHEMA, json.toString());

    if (obj.has(RESULT)) {
      return context.deserialize(json, Result.class);
    } else if (obj.has(ERROR)) {
      return context.deserialize(json, Error.class);
    } else {
      throw new JsonParseException("A result must contain one of the field result or error");
    }
  }

  @Override
  public JsonElement serialize(Answer src, Type typeOfSrc, JsonSerializationContext context) {
    JsonObject obj = context.serialize(src).getAsJsonObject();
    obj.addProperty(JsonUtils.JSON_RPC, JsonUtils.JSON_RPC_VERSION);
    JsonUtils.verifyJson(JsonUtils.ROOT_SCHEMA, obj.toString());
    return obj;
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/network/JsonAnswerSerializer.kt
package com.github.dedis.popstellar.model.network.serializer.network

import com.github.dedis.popstellar.model.network.answer.Answer
import com.github.dedis.popstellar.model.network.answer.Error
import com.github.dedis.popstellar.model.network.answer.Result
import com.github.dedis.popstellar.model.network.serializer.JsonUtils
import com.github.dedis.popstellar.model.network.serializer.JsonUtils.testRPCVersion
import com.github.dedis.popstellar.model.network.serializer.JsonUtils.verifyJson
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.lang.reflect.Type

/** Json serializer and deserializer for the answer */
class JsonAnswerSerializer : JsonSerializer<Answer?>, JsonDeserializer<Answer> {
  @Throws(JsonParseException::class)
  override fun deserialize(
    json: JsonElement,
    typeOfT: Type,
    context: JsonDeserializationContext
  ): Answer {
    val obj = json.asJsonObject

    testRPCVersion(obj)
    verifyJson(JsonUtils.ROOT_SCHEMA, json.toString())

    return if (obj.has(RESULT)) {
      context.deserialize(json, Result::class.java)
    } else if (obj.has(ERROR)) {
      context.deserialize<Answer>(json, Error::class.java)
    } else {
      throw JsonParseException("A result must contain one of the field result or error")
    }
  }

  override fun serialize(
    src: Answer?,
    typeOfSrc: Type,
    context: JsonSerializationContext
  ): JsonElement {
    val obj = context.serialize(src).asJsonObject
    obj.addProperty(JsonUtils.JSON_RPC, JsonUtils.JSON_RPC_VERSION)

    verifyJson(JsonUtils.ROOT_SCHEMA, obj.toString())

    return obj
  }

  companion object {
    private const val RESULT = "result"
    private const val ERROR = "error"
  }
}
